src/rdkafka_fetcher.c

```diff
@@ -959,6 +1448,483 @@ static rd_bool_t can_use_topic_ids(rd_kafka_broker_t *rkb) {
+                                rd_kafka_buf_write_i8(rkbuf, 1); /* ACCEPT */
+                                /* Acknowledgement tags */
+                                rd_kafka_buf_write_tags_empty(rkbuf);
+                        }
+                        rktp->rktp_share_acknowledge_count = 0;
+                        rd_free(rktp->rktp_share_acknowledge);
+                        rktp->rktp_share_acknowledge = NULL;
+                } else {
+                        rd_rkb_dbg(rkb, FETCH, "SHAREFETCH", "rd_kafka_ShareFetchRequest: topic %.*s [%" PRId32 "] : No acknowledgements to send",
+                                RD_KAFKAP_STR_PR(rktp->rktp_rkt->rkt_topic),
+                                rktp->rktp_partition);
+                        /* No acknowledgements */
+                        rd_kafka_buf_write_arraycnt(rkbuf, 0);
+                }
+
+                /* Partition tags */
+                rd_kafka_buf_write_tags_empty(rkbuf);
+
+                rd_rkb_dbg(rkb, FETCH, "SHAREFETCH",
+                        "Share Fetch topic %.*s [%" PRId32 "]",
+                        RD_KAFKAP_STR_PR(rktp->rktp_rkt->rkt_topic),
+                        rktp->rktp_partition);
+
+                cnt++;
+        }
+
+        rd_kafka_dbg(rkb->rkb_rk, FETCH, "SHAREFETCH",
+                   "Share Fetch Request with %d toppars on %d topics",
+                   cnt, TopicArrayCnt);
+
+        if (rkt_last != NULL) {
+                /* Update last topic's PartitionArrayCnt */
+                rd_kafka_buf_finalize_arraycnt(rkbuf, of_PartitionArrayCnt,
+                                               PartitionArrayCnt);
+                /* Topic tags */
+                rd_kafka_buf_write_tags_empty(rkbuf);
+        }
+
+        /* Update TopicArrayCnt */
+        rd_kafka_buf_finalize_arraycnt(rkbuf, of_TopicArrayCnt, TopicArrayCnt);
+
+        if(toppars_to_send) {
+                rd_list_destroy(toppars_to_send);
+        }
+
+        /*
+         * TODO KIP-932: Move this to the caller.
+         */
+        if(is_leave_request || has_acknowledgements_or_topics_to_add || has_toppars_to_forget || is_fetching_messages) {
+                rd_kafka_dbg(rkb->rkb_rk, FETCH, "SHAREFETCH",
+                           "Share Fetch Request sent with%s%s%s",
+                           has_acknowledgements_or_topics_to_add ? " acknowledgements," : "",
+                           has_toppars_to_forget ? " forgotten toppars," : "",
+                           is_fetching_messages ? " fetching messages" : "");
+        } else {
+                rd_kafka_buf_destroy(rkbuf);
+                rd_kafka_dbg(rkb->rkb_rk, FETCH, "SHAREFETCH",
+                           "Share Fetch Request not sent since there are no "
+                           "acknowledgements, forgotten toppars or messages to fetch");
+                rd_kafka_op_reply(rko_orig, RD_KAFKA_RESP_ERR__NOOP);
+                return;
+        }
```

Comment:
Move this to the caller as we know if there are acknowledgement or not before calling this function itself.